### DISCUSSION TITLE
Boundary condition depending on the State

### DISCUSSION THREAD

**Discussion by Waschenbacher** [zwicker-group/py-pde#488] [Q&A] [answered]
I am implementing a custom PDE class and would like to add a not-so-easy boundary condition. Is it possible in py-PDE to add a Dirichlet boundary condition with its value being a function of 
- its fields.data at coordinate (x = L) 
- f(t)
Example: I have a field A on 1-d "grid = CartesianGrid([[0, 10]], [100])]". The boundary condition at left-hand-side should be a function of the field at the right and a function of time t.

**Comment by david-zwicker** [comment]
If I understand you correctly, you would like to have a custom periodic boundary condition. This should in principle be possible, although the condition would not depend on the exact value at the boundary but rather on the value of the support point next to the boundary (a distance `dx/2` away from the boundary). Otherwise, the situation cannot be resolved unambiguously. Would this cover your scenario? Obviously, you'd also need a boundary condition on the right side, but I guess this could be whatever you want it to be.

Such conditions are currently not implemented, but this should be fairly straightforward. However, I would need to have a good term for such a condition. Do you have a suggestion?

**Reply by Waschenbacher** [reply]
Thank you very much for the quick reply. I'm not sure whether I understand you correctly about the "custom periodic boundary condition". In my scenario, I would like to simulate multiple ingredients flowing through a pipe. Here, the inlet is controlled by a feedback controller, which depends on the ingredients at the outlet. To simulate this closed loop problem, 
- I use a multi-dimensional convection-diffusion model to model the concentration levels inside the pipe.
- The left boundary condition is controlled by a controller, which depends on the outlet of the pipe and some other custom variables changing over time. 

I believe what helps in my situation is a more flexible boundary condition, which should be a custom function of the values at a specific coordinate(in my scenario the right boundary) and several other custom variables.

For this simulation, in the past I used a PDE solver (Finite-Volumn-method), which runs the situation on discrete time steps. I'm hoping to speed up the simulation by using the method-of-lines. But I didn't manage to set the boundary conditions.

**Reply by david-zwicker** [reply]
I just merged #490, which allows more control over the value that determine the boundary condition. Using the latest `master` branch, you should be able to use a boundary condition like `{"type": "value_expression", "value": "sin(value)", "value_cell": -1}`, which would set the value of the BC to the sine of the value of the right-most grid cell. Let me know if this covers your use case!

**Reply by Waschenbacher** [reply]
Awesome! Thanks! It worked with sin. However, I am wondering how can I use a user-defined-func instead of sin to this boundary condition. According to the Expression section from https://py-pde.readthedocs.io/en/latest/manual/advanced_usage.html#boundary-conditions, I tried out `{"type": "value_expression", "user_funcs": "user_defined_boundary(value)", "value_cell": -1}` with user_defined_boundary being the user defined function, but it seems that the BC does not accept user_funcs.

**Reply by david-zwicker** [reply]
Good point. This should be easy to fix. I'll open another issue and hope to work on this soon.

**Reply by david-zwicker** [reply, marked as answer]
I just pushed #492 , which should allow you to use `{"type": "value_expression", "value": "func(value)", "value_cell": -1, "user_funcs": {"func": user_defined_boundary}}`. I hope this is what you meant.

**Reply by Waschenbacher** [reply]
thank you so much for the new feature! I really appreciate it a lot! It works with a general user_funcs. However when I tried it with the following case: 
`class PID:

    def __init__(self, Kp, Ki, Kd, setpoint):
        self.Kp = Kp
        self.Ki = Ki
        self.Kd = Kd
        self.setpoint = setpoint

    def get_ctrl_func(self, measurement):
        ...
        return ...
    
`pid = PID(Kp=1., Ki=0., Kd=0., setpoint=1.)

def user_boundary(x: float) -> float:
    return pid.get_ctrl_func(x)`


I received the error message below:
<img width="1214" alt="Screenshot 2023-12-14 at 23 57 10" src="https://github.com/zwicker-group/py-pde/assets/116073761/d56f69ac-8dba-4a70-bc33-701616ed4712">
Could you please give me some further guidance?

**Reply by david-zwicker** [reply]
The error indicates that the user-supplied function cannot be compiled using `numba`. I guess it would be useful if `py-pde` then falls back onto the basic python backend, which would be much slower, but could support more python code. Could you please raise an issue with reference to this discussion, so we can improve this in the future?

**Reply by Waschenbacher** [reply]
Thank! I just opened an issue

**Reply by david-zwicker** [reply]
I found some time and implemented a change in the latest `master` branch that would allow calling a non-compilable function. I hope it's helpful!

**Reply by david-zwicker** [reply]
I fixed the last implementation so it hopefully now works as intended. Please reopen the issue if there are still problems.
